File: src/main/java/fr/ribesg/blob/command/util/UrbanCommand.java
package fr.ribesg.blob.command.util;

import com.google.gson.*;
import fr.ribesg.alix.api.*;
import fr.ribesg.alix.api.bot.command.Command;
import fr.ribesg.alix.api.bot.util.WebUtil;
import fr.ribesg.alix.api.enums.Codes;
import org.apache.commons.lang3.StringUtils;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

public class UrbanCommand extends Command {

   public UrbanCommand() {
      super("urban", new String[]{
         "Find something on UrbanDictionary",
         "Usage: ##[.number] <query>"
      }, "u");
   }

   @Override
   public boolean exec(final Server server, final Channel channel, final Source user, final String primaryArgument, final String[] args) {
      final Receiver receiver = channel == null ? user : channel;

      if (args.length == 0) {
         return false;
      }

      int definitionNumber;
      try {
         definitionNumber = Integer.parseInt(primaryArgument);
         if (definitionNumber < 1) {
            definitionNumber = 1;
         }
      } catch (final NumberFormatException e) {
         definitionNumber = 1;
      }
      final String query = StringUtils.join(args, " ");
      final String url;
      try {
         url = "http://urbanscraper.herokuapp.com/search/" + URLEncoder.encode(query, "UTF-8");
      } catch (final UnsupportedEncodingException e) {
         Log.error(e.getMessage(), e);
         receiver.sendMessage(Codes.RED + (channel == null ? "" : user.getName() + ", ") + "failed to encode that thing to a valid URL parameter!");
         return true;
      }

      final String response;
      try {
         response = WebUtil.get(url);
      } catch (final IOException e) {
         Log.error(e.getMessage(), e);
         receiver.sendMessage(Codes.RED + (channel == null ? "" : user.getName() + ", ") + "failed to get result from API!");
         return true;
      }

      final JsonElement jsonResponse;
      try {
         jsonResponse = new JsonParser().parse(response);
      } catch (final JsonParseException e) {
         Log.error(e.getMessage(), e);
         receiver.sendMessage(Codes.RED + (channel == null ? "" : user.getName() + ", ") + "failed to parse response from API!");
         return true;
      }

      if (jsonResponse.isJsonArray()) {
         final JsonArray array = jsonResponse.getAsJsonArray();
         final int size = array.size();
         if (size == 0) {
            receiver.sendMessage(Codes.RED + (channel == null ? "" : user.getName() + ", ") + "no result.");
         } else if (definitionNumber > size) {
            receiver.sendMessage(Codes.RED + (channel == null ? "" : user.getName() + ", ") + "no such definition number.");
         } else {
            final JsonElement definitionElement = array.get(definitionNumber - 1);
            if (definitionElement.isJsonObject()) {
               final JsonObject definitionObject = definitionElement.getAsJsonObject();
               if (definitionObject.has("definition")) {
                  String definitionString = definitionObject.getAsJsonPrimitive("definition").getAsString();
                  if (definitionString.length() > 200) {
                     definitionString = definitionString.substring(0, 197) + "...";
                  }
                  final String definitionUrl;
                  if (definitionObject.has("url")) {
                     definitionUrl = definitionObject.getAsJsonPrimitive("url").getAsString();
                     String shortUrl;
                     try {
                        shortUrl = WebUtil.shortenUrl(definitionUrl);
                     } catch (IOException e) {
                        Log.error("Failed to shorten URL '" + definitionUrl + "'", e);
                        shortUrl = definitionUrl;
                     }
                     receiver.sendMessage((channel == null ? "" : user.getName() + ", ") + definitionString + '(' + shortUrl + ')');
                  } else {
                     receiver.sendMessage((channel == null ? "" : user.getName() + ", ") + ", " + definitionString);
                  }
               } else {
                  Log.error("Malformed response from API!");
                  receiver.sendMessage(Codes.RED + (channel == null ? "" : user.getName() + ", ") + "malformed response from API!");
               }
            }
         }
      }
      return true;
   }
}

File: src/main/kotlin/fr/ribesg/blob/command/util/UrbanCommand.kt
package fr.ribesg.blob.command.util

import com.google.gson.JsonParseException
import com.google.gson.JsonParser
import fr.ribesg.blob.command.Command
import fr.ribesg.blob.extensions.sendRedNotice
import fr.ribesg.blob.util.Log
import fr.ribesg.blob.util.WebUtil
import org.kitteh.irc.client.library.Client
import org.kitteh.irc.client.library.element.Channel
import org.kitteh.irc.client.library.element.User
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URLEncoder

/**
 * @author Ribesg
 */

public class UrbanCommand(prefix: Char) : Command(prefix) {

    init {
        this.usage = this.prefix + "urban <query>"
    }

    override fun exec(client: Client, from: User, to: Channel?, primArg: String?, args: Array<String>) {
        if (args.size() == 0) {
            sendUsage(from)
        } else {

            val definitionNumber: Int = try {
                var tmp = Integer.parseInt(primArg!!)
                if (tmp < 1) 1 else tmp
            } catch (e: Exception) {
                1
            }

            val query = args.join(" ")
            val url = try {
                "http://urbanscraper.herokuapp.com/search/" + URLEncoder.encode(query, "UTF-8")
            } catch (e: UnsupportedEncodingException) {
                // Should not happen
                Log.error(e.getMessage() ?: "Error", e)
                return
            }

            val response = try {
                WebUtil.get(url)
            } catch (e: IOException) {
                from.sendRedNotice("Failed to get result from UrbanDictionary API!")
                Log.error(e.getMessage() ?: "Error", e)
                return
            }

            val jsonResponse = try {
                JsonParser().parse(response)
            } catch (e: JsonParseException) {
                from.sendRedNotice("Failed to parse response from UrbanDictionary API!")
                Log.error(e.getMessage() ?: "Error", e)
                return
            }

            if (!jsonResponse.isJsonArray()) return

            val array = jsonResponse.getAsJsonArray()
            val size = array.size()
            if (definitionNumber > size) {
                if (size == 0) {
                    (to ?: from).sendMessage(from.getNick() + ", no results")
                } else {
                    (to ?: from).sendMessage(from.getNick() + ", no such definition number")
                }
                return
            }

            val definitionElement = array.get(definitionNumber - 1);
            if (definitionElement.isJsonObject()) {
                val definitionObject = definitionElement.getAsJsonObject();
                if (definitionObject.has("definition")) {
                    var definitionString = definitionObject.getAsJsonPrimitive("definition").getAsString()
                    if (definitionString.length() > 200) {
                        definitionString = definitionString.substring(0, 197) + "..."
                    }
                    if (definitionObject.has("url")) {
                        val definitionUrl = definitionObject.getAsJsonPrimitive("url").getAsString()
                        val shortUrl = try {
                            WebUtil.shortenUrl(definitionUrl)
                        } catch (e: IOException) {
                            Log.error("Failed to shorten URL '$definitionUrl'", e)
                            definitionUrl
                        }
                        (to ?: from).sendMessage(from.getNick() + ", $definitionString ($shortUrl)")
                    } else {
                        (to ?: from).sendMessage(from.getNick() + ", $definitionString")
                    }
                } else {
                    from.sendRedNotice("Malformed response from API!")
                    Log.error("Malformed response from API!")
                }
            }
        }
    }
}
